# **MÓDULO 13**
# Projeto - Fundamentos da Descoberta de Dados

Nesse projeto trabalharemos com a base de dados de produtos de um supermercado do Chile.
A ideia é que vocês apliquem os conceitos estatísticos vistos no último módulo, mais os conceitos de visualizações de dados através de gráficos e finalizem publicando no seu github!

In [14]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

Faça a leitura dos dados do arquivo CSV:


Altere o código abaixo de acordo com seu diretório.

In [127]:
df = pd.read_csv("MODULO7_PROJETOFINAL_BASE_SUPERMERCADO.csv", delimiter=',')

df_filtered = df.loc[df['Categoria'] == 'lacteos']

df_filtered_by_price = df_filtered[df_filtered['Preco_Normal'] > 5934]

print(f"Total: {len(df)} / Filtered by Lacteos: {len(df_filtered)} / By Lacteos and Price: {len(df_filtered_by_price)}")

# df_filtered['Preco_Normal'].sum()
# df_filtered['Preco_Normal'].describe()
# df['Preco_Normal'].count()
# df.groupby('Categoria')['Preco_Normal'].count()
# df_filtered.head(100)

df.head(10)

Total: 1107 / Filtered by Lacteos: 447 / By Lacteos and Price: 43


title       Marca  Preco_Normal  \
0              Pack 12 un, Leche extra proteína 1 L  Loncoleche         19788   
1   Pack 12 un, Leche chocolate receta original 1 L     Soprole         18228   
2    Pack 12 un, Leche semidescremada chocolate 1 L     Soprole         18228   
3     Pack 12 un, Leche semidescremada frutilla 1 L     Soprole         18228   
4       Pack 12 un, Leche sin lactosa chocolate 1 L  Loncoleche         17988   
5        Pack 12 un, Leche sin lactosa frutilla 1 L  Loncoleche         17988   
6  Pack 12 un, Leche saborizada light chocolate 1 L  Loncoleche         17988   
7         Pack 12 un, Leche saborizada frutilla 1 L       Colun         17388   
8         Pack 12 un, Leche saborizada vainilla 1 L       Colun         17388   
9           Pack 12 un, Leche saborizada manjar 1 L       Colun         17388   

   Preco_Desconto  Preco_Anterior  Desconto Categoria  
0               0               0         0   lacteos  
1               0               0         0   lacteos  
2               0               0         0   lacteos  
3               0               0         0   lacteos  
4               0               0         0   lacteos  
5               0               0         0   lacteos  
6               0               0         0   lacteos  
7               0               0         0   lacteos  
8               0               0         0   lacteos  
9               0               0         0   lacteos

Os campos do nosso dataframe são:

**Title:** Nome do produto.


**Marca:** A marca do produto.


**Preco_Normal:** O preço em que o produto costuma ser vendido quando não há desconto.


**Preco_Desconto:** O preço vendido após o desconto ser aplicado.


**Preco_Anterior:** Preço em que era comercializado o produto antes do desconto aplicado.


**Desconto:** Total de desconto aplicado.






As colunas que aparecem com valores 0 são para os produtos onde não tivemos descontos aplicados.


As categorias estão em espanhol!

# 1 - Traga a média e a mediana dos preços - coluna Preco_Normal - por categoria de produto.
# Identifique as categorias que parecem ter um valor de média abaixo ou acima da mediana.

In [41]:
# Seu código aqui para a média
df.groupby('Categoria')['Preco_Normal'].mean().reset_index().sort_values(by='Preco_Normal', ascending=True)

Categoria  Preco_Normal
4        instantaneos-y-sopas    765.491228
6                    verduras   1343.296875
3                      frutas   1724.473684
0  belleza-y-cuidado-personal   1783.556485
2                  congelados   2108.042553
5                     lacteos   2385.219239
1          comidas-preparadas   3095.043478

In [42]:
# Seu código aqui para a mediana
df.groupby('Categoria')['Preco_Normal'].median().reset_index().sort_values(by='Preco_Normal', ascending=True)

Categoria  Preco_Normal
4        instantaneos-y-sopas         439.0
5                     lacteos         989.0
6                    verduras        1180.0
3                      frutas        1195.0
2                  congelados        1519.0
0  belleza-y-cuidado-personal        1569.0
1          comidas-preparadas        3290.0

### Valores de média **abaixo** da mediana:
- comidas-preparadas

### Valores de média **acima** da mediana:
- instantaneos-y-sopas
- verduras
- frutas
- congelados
- lacteos
- belleza-y-cuidado-personal


# 2 - Traga o desvio padrão por categoria de produto.
# Qual o comportamento da média e mediana nas categorias com maior desvio?

In [52]:
# Seu código aqui
df.groupby('Categoria')['Preco_Normal'].std().reset_index().sort_values(by='Preco_Normal', ascending=False)

# Desvio padrão central
# df['Preco_Normal'].std() # 2961.25762666094

# >>> Média

# 5	lacteos	2385.219239
# 0	belleza-y-cuidado-personal	1783.556485
# 2	congelados	2108.042553
# 1	comidas-preparadas	3095.043478
# 3	frutas	1724.473684
# 4	instantaneos-y-sopas	765.491228
# 6	verduras	1343.296875

# >>> Mediana

# 5	lacteos	989.0
# 0	belleza-y-cuidado-personal	1569.0
# 2	congelados	1519.0
# 1	comidas-preparadas	3290.0
# 3	frutas	1195.0
# 4	instantaneos-y-sopas	439.0
# 6	verduras	1180.0

# >>> Desvio Padrão

# 5	lacteos	3925.816164
# 0	belleza-y-cuidado-personal	2210.041719
# 2	congelados	2111.539896
# 1	comidas-preparadas	2019.911428
# 3	frutas	1639.151114
# 4	instantaneos-y-sopas	1170.232869
# 6	verduras	1012.699625

# >>> Conclusões

# Categoria "lacteos"
# 989.0 (Mediana mais distante) > 2385.2 (Média mais próximo do DP) > 3925.8 (Desvio Padrão)

# Categoria "belleza-y-cuidado-personal"
# 1569.0 (Mediana mais distante) > 1783.5 (Média mais próximo do DP) > 2210.0 (Desvio Padrão)

# Categoria "congelados"
# 1519.0 (Mediana mais distante) > 2108.0 (Média) > 2111.5 (Desvio Padrão)

Categoria  Preco_Normal
5                     lacteos   3925.816164
0  belleza-y-cuidado-personal   2210.041719
2                  congelados   2111.539896
1          comidas-preparadas   2019.911428
3                      frutas   1639.151114
4        instantaneos-y-sopas   1170.232869
6                    verduras   1012.699625

### Digite nesse campo o comportamento que você identificou.

As três categorias com maior desvio padrão são a lacteos, belleza-y-cuidado-personal e congelados. Elas possuem uma mediana bem discrepante do desvio padrão, mas possuem uma média bem próxima do desvio padrão.

# 3 - Plot um boxplot da distribuição do Preco_Normal para a categoria que você identificou que tem o maior desvio padrão. Como é a distribuição desses dados segundo o boxplot? Você identifica muitos outliers?

### Distribuição dos dados

- Mínimo em 0
- Q1 em 269
- Mediana (Q2) em 989
- Q3 em 2619
- Máximo em 5934 (valor de corte para encontrar os outliers acima disso)

### Outliers

Sim, podemos identificar uma série de outliers presente nos dados (exatos 43 registros dentre os 447 da categoria "lacteos"). Para identificá-lo, basta olharmos para o valor máximo definido pelo boxplot e filtrar os dados pela coluna "Preco_Normal" no dataframe que passar disso.

Dica: Para trazer apenas os dados da categoria que você deseja você pode usar o df.loc[df['Categoria'] == 'CATEGORIA ESCOLHIDA'

In [93]:
# Seu código aqui

filtered_data = df.loc[df['Categoria'] == 'lacteos']
fig = px.box(filtered_data, y='Preco_Normal',
             title='Distribuição dos Preços dos produtos - Categoria Lacteos',
             points='all',
             labels={'Preco_Normal': "Preço Normal"},
             notched=True
            )

fig.show()

# 4 - Plote um gráfico de barras onde temos a média de descontos por categoria.

In [109]:
# Seu código aqui

media_descontos_por_categoria = df.groupby('Categoria')['Desconto'].mean().reset_index().sort_values(by='Desconto', ascending=True)
fig = px.bar(media_descontos_por_categoria,
             x='Desconto',
             y='Categoria',
             title='Média de descontos por categoria',
             labels={'Desconto': "Média de Descontos", 'Categoria': "Categoria"},
             color='Categoria',
             text='Desconto'
            )

fig.update_traces(texttemplate='%{value:0.2f}')
fig.show()

# 5 - Plote um gráfico de mapa interativo agrupando os dados por categoria, marca e trazendo a média de desconto.

In [126]:
# Seu código aqui
data = df.groupby(['Categoria', 'Marca'])['Desconto'].mean().reset_index()
fig = px.treemap(data,
                 path=['Categoria', 'Marca'],
                 values='Desconto',
                 color='Categoria'
                )
fig.update_traces(root_color="lightgrey")
fig.update_layout(margin = dict(t=50, l=50, r=50, b=50))
fig.update_traces(marker=dict(cornerradius=5))
fig.show()